hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/MiniOzoneClusterImpl.java

```diff
@@ -531,16 +511,18 @@ private static void stopRecon(ReconServer reconServer) {
     }
   }
 
-  private static void stopS3G(Gateway s3g) {
-    try {
-      if (s3g != null) {
-        LOG.info("Stopping S3G");
-        // TODO (HDDS-11539): Remove this workaround once the @PreDestroy issue is fixed
-        OzoneClientCache.closeClient();
-        s3g.stop();
+  private static void stopServices(List<Service> services) {
+    // stop in reverse order
+    List<Service> reverse = new ArrayList<>(services);
+    Collections.reverse(reverse);
+
+    for (Service service : reverse) {
+      try {
+        service.stop();
+        LOG.info("Stopped {}", service);
+      } catch (Exception e) {
+        LOG.error("Error stopping {}", service, e);
```

Comment:
nit:
```suggestion
    ListIterator<Service> iterator = services.listIterator(services.size());
    while (iterator.hasPrevious()) {
      Service service = iterator.previous();
      try {
        service.stop();
        LOG.info("Stopped {}", service);
      } catch (Exception e) {
        LOG.error("Error stopping {}", service, e);
      }
    }
```
However, this has worse readability, so please take it with a grain of salt.

Reply:
Thanks for the suggestion.  I'll keep the original, because I don't expect many services (only Recon (HDDS-12568) and maybe S3G HA (HDDS-11679)).

This is just a precaution to allow dependent services to be added in dependency order.